src/main/scala/analysis/Local.scala

```diff
@@ -0,0 +1,337 @@
+package analysis
+
+import analysis.BitVectorEval.{bv2SignedInt, isNegative}
+import ir.{Assign, BVADD, BinaryExpr, BitVecLiteral, BitVecType, CFGPosition, DirectCall, Endian, Expr, Extract, IntraProcIRCursor, MemoryAssign, MemoryLoad, Procedure, Register, Variable, ZeroExtend, computeDomain, toShortString}
+import specification.{ExternalFunction, SpecGlobal, SymbolTableEntry}
+
+import scala.util.control.Breaks.{break, breakable}
+import java.math.BigInteger
+import scala.collection.mutable
+
+/**
+ * The local phase of Data Structure Analysis
+ * @param proc procedure to be analysed
+ * @param symResults result of symbolic access analysis
+ * @param constProp
+ * @param globals
+ * @param globalOffsets
+ * @param externalFunctions
+ * @param reachingDefs
+ * @param writesTo mapping from procedures to registers they change
+ * @param params mapping from procedures to their parameters
+ */
+class Local(
+             proc: Procedure,
+             symResults: Map[CFGPosition, Map[SymbolicAccess, TwoElement]],
+             constProp: Map[CFGPosition, Map[Variable, FlatElement[BitVecLiteral]]],
+             globals: Set[SymbolTableEntry], globalOffsets: Map[BigInt, BigInt],
+             externalFunctions: Set[ExternalFunction],
+             reachingDefs: Map[CFGPosition, Map[Variable, Set[CFGPosition]]],
+             writesTo: Map[Procedure, Set[Register]],
+             params:  Map[Procedure, Set[Variable]]
+           ) extends Analysis[Any]{
+
+  private val mallocRegister = Register("R0", 64)
+  private val stackPointer = Register("R31", 64)
+
+  // set of cfg positions already processed by the analysis local phase
+  private val visited: mutable.Set[CFGPosition] = mutable.Set()
+
+
+  // variables to symbolic access map for each cfg position
+  val varToSym: Map[CFGPosition, Map[Variable, Set[SymbolicAccess]]] = symResults.foldLeft(Map[CFGPosition, Map[Variable, Set[SymbolicAccess]]]()) {
+    (outerMap, syms) =>
+      val position = syms._1
+      val innerMap = syms._2.foldLeft(Map[Variable, Set[SymbolicAccess]]()) {
+        (m, access) =>
+          if m.contains(access._1.accessor) then
+            // every variable pointing to a stack region ONLY has one symbolic access associated with it.
+            m(access._1.accessor).foreach(
+              sym => assert(!sym.symbolicBase.isInstanceOf[StackLocation])
```

Comment:
This assertion fails for a number of test cases when run with the analysis on:
- correct/indirect_call/clang:GTIRB
- correct/indirect_call/clang_no_plt_no_pic:GTIRB
- correct/indirect_call/clang_pic:GTIRB
- correct/indirect_call/gcc:GTIRB
- correct/indirect_call/gcc_no_plt_no_pic:GTIRB
- correct/indirect_call/gcc_pic:GTIRB
- correct/malloc_with_local3/clang:GTIRB
- correct/malloc_with_local3/clang_no_plt_no_pic:GTIRB
- correct/malloc_with_local3/clang_pic:GTIRB
- correct/malloc_with_local3/gcc:GTIRB
- correct/malloc_with_local3/gcc_no_plt_no_pic:GTIRB
- correct/malloc_with_local3/gcc_pic:GTIRB

Reply:
I think an issue here are direct calls to procedures that don't have a return block. The ide solver treats these as branches instead of procedure calls. the function entry will be the join of all the direct calls for this procedure. therefore there will be multiple (one for each caller) stack locations associated with R31 in the callee. The simplest way around it is to not propagate stack symbolic addresses through direct calls (this already takes place for direct calls which the ide solver recognizes as a call (ones with a return block)). Otherwise will have to merge stack memory regions of different procedures together. (this is different to merging of stack elements in the same procedures stack which 

I have implemented the former and ran systemtests and the assertion no longer fails. I still think that the assertion may fail in some rare cases but not for any variations of the examples so far. I'm going to try an construct an example where it fails and work on it from there

Reply:
There are definitely other cases where the assertions can fail. See this case: https://github.com/UQ-PAC/BASIL/tree/points-to-analysis-alternate-merge-stackambiguity/examples/stackambiguity

Reply:
In this example, R9 and R10 are assigned different stack pointer offsets depending on branch conditions, which causes them to have multiple `SymbolicAddress`s associated with them after the branching.

Reply:
refactored the stack recognition functionality to allow for  multiple stack symbolic addresses associated with a register. It first finds all the stack offsets and their maximum accessed size. It then builds the stack regions by merging overlapping accesses to a single Node (each offset in the overlapping access will be a cell of this node).
Offsets which don't have any overlapping accesses remain as separate nodes

Reply:
okay i think there is still the possibility of an indirect overlapping access though. ie store stack address in heap, load it, use loaded address to do a 128 bit load. Not sure how to handle such cases. Ian iterative approach where we find indirect stack references and redo the analysis?

Reply:
That seems like something that should be handled during the DSA itself - you need to deal with it for non-stack accesses as well. You need some sort of check to find if an access is accessing multiple nodes (at once) that you would then concatenate & merge.